### PR TITLE
fix: fresh objects for Synchronization executions

### PR DESCRIPTION
**Pull request by diafour**
#### Overview

Set `objects` in "Synchronization" binding context to an actual state in cluster.

#### What this PR does / why we need it

- snapshots in monitors are updated without Event tasks until success Synchronization execution
- fix documentation for "Group" binding context

#### Special notes for your reviewer

Integration tests are fixed, but there are 3 clusters in use. We need to refactor this part in the future.

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/hook/controller/kubernetes_bindings_controller.go [hunk 1]
```diff
@@ -50,6 +47,10 @@ type kubernetesBindingsController struct {
 
 	// dependencies
 	kubeEventsManager kube_events_manager.KubeEventsManager
+
+	// Snapshots cache for UpdateSnapshots
+	snapshotsCache      map[string][]ObjectAndFilterResult
```
Comment (nabokihms) [inline pkg/hook/controller/kubernetes_bindings_controller.go:52]: Why do we need a separate cache for snapshots? We already have a cache for filtered objects in the resource informer. Frankly, I do not understand the whole part about cached snapshots mode 😅  Cant we enable this mode by default without deferring and all staff?

File: HOOKS.md [hunk 2]
```diff
@@ -693,13 +693,13 @@ kubernetes:
 
 #### "Synchronization" binding context for group
 
-During startup, the hook will be executed with the "Synchronization" binding context with `snapshots` JSON object:
+During startup, the hook will be executed with the "Group" binding context with `snapshots` JSON object that contains all objects from bindings in the group. Note there is no `objects` field as for "normal" binding:
```
[the remark this answers is not in the document]
Comment (diafour) [reply HOOKS.md:696]: I've rewrote this section completely.

File: pkg/hook/controller/hook_controller.go [hunk 3]
```diff
@@ -306,10 +314,17 @@ func (hc *hookController) UpdateSnapshots(context []BindingContext) []BindingCon
...
+	// Turn on cached snapshots mode to retrieve snapshots only once. (Synchronization and self-include)
+	hc.KubernetesController.StartCachedSnapshotMode()
+	defer hc.KubernetesController.StopCachedSnapshotMode()
+
 	newContext := []BindingContext{}
 	for _, bc := range context {
 		newBc := bc
 		newBc.Snapshots = hc.KubernetesSnapshotsFor(bc.Metadata.BindingType, bc.Binding)
+		if newBc.Metadata.BindingType == OnKubernetesEvent && newBc.Type == TypeSynchronization {
+			newBc.Objects = hc.KubernetesController.SnapshotsFor(bc.Binding)
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/hook/controller/hook_controller.go:326]: `newBC.Objects` is just the holder for these two variants:

"objects" will contain an actual state for the moment of the hook execution. This is an array of existed objects, and in the case of error, _it should be updated_ for the next execution.

"object" in "Event" binding context is the state for the moment of the event. It should not be updated. Actual object state is in the snapshot.

I've tried to explain this more clear in HOOKS.md.

File: pkg/shell-operator/operator.go [hunk 4]
```diff
@@ -673,74 +746,56 @@ func (op *ShellOperator) TaskHandleHookRun(t task.Task) queue.TaskResult {
...
-			"hook": hookMeta.HookName,
-		})
+		err = op.ObjectPatcher.GenerateFromJSONAndExecuteOperations(specs)
 		if err != nil {
-			goto Metrics
+			return err
 		}
 	}
 
-Metrics:
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/shell-operator/operator.go:709]: Hook execution for the "HooRun" task is optional now, so I've extracted the hook execution logic into `HandleRunHook` method for simplicity. As a side effect, the unusual chaining of `err` checks is replaced with the usual return.

File: pkg/hook/controller/kubernetes_bindings_controller.go (hunk 1 above)
Comment (diafour) [reply pkg/hook/controller/kubernetes_bindings_controller.go:52]: The "cached snapshots mode" is used only for the "Synchronization" phase. 

Binding context preparation for combined "Synchronization" binging contexts or for "Synchronization" with self-inclusion may require several calls to Snapshot() and objects can change between these calls. So the cache is introduced to ensure snapshot consistency.

File: pkg/kube_events_manager/kube_events_manager_test.go [hunk 5]
```diff
@@ -336,12 +336,12 @@ func Test_FakeClient_CatchUpdates(t *testing.T) {
...
 	monitor.Metadata.MonitorId = "MonitorId"
 
-	_, err := mgr.AddMonitor(monitor)
+	err := mgr.AddMonitor(monitor)
 	if !assert.NoError(t, err) {
 		t.FailNow()
 	}
 
-	mgr.Start()
+	//mgr.Start()
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/kube_events_manager/kube_events_manager_test.go:344]: There is no Start method anymore, I've deleted this line.

File: pkg/kube_events_manager/resource_informer.go [hunk 6]
```diff
@@ -158,21 +162,39 @@ func (ei *resourceInformer) CreateSharedInformer() (err error) {
...
 	ei.cacheLock.RLock()
-	defer ei.cacheLock.RUnlock()
 	res := make([]ObjectAndFilterResult, 0)
-	for _, obj := range ei.CachedObjects {
+	for _, obj := range ei.cachedObjects {
 		res = append(res, *obj)
 	}
+	ei.cacheLock.RUnlock()
+	// Reset event buf.
+	ei.cacheLock.Lock()
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/kube_events_manager/resource_informer.go:174]: An `eventBufLock` is introduced.

File: pkg/kube_events_manager/resource_informer.go [hunk 7]
```diff
@@ -299,13 +321,15 @@ func (ei *resourceInformer) HandleWatchEvent(object interface{}, eventType Watch
 
 	// Do not fire Added or Modified if object is in cache and its checksum is equal to the newChecksum.
 	// Delete is always fired.
+	ei.cacheLock.Lock()
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/kube_events_manager/resource_informer.go:324]: This one is reverted.

File: HOOKS.md [hunk 8]
```diff
@@ -659,49 +674,67 @@ at 12:02, the hook will be executed with the following binding context:
 
 ### Binding context of grouped bindings
 
-`group` parameter defines a named group of bindings. Group is used when the source of event is not important and data in snapshots is enough for the hook. When binding with `group` is triggered with the event, the hook receives snapshots from all bindings with equal `group` name. Also, adjacent tasks with equal `group` in the same queue are "compacted" and hook is executed only once. So it is wise to use the same queue for all hooks in a group.
+`group` parameter defines a named group of bindings. Group is used when the source of event is not important and data in snapshots is enough for the hook. When binding with `group` is triggered with the event, the hook receives snapshots from all `kubernetes` bindings with the same `group` name.
```
Comment (nabokihms) [inline HOOKS.md:677]:
```suggestion
`group` parameter defines a named group of bindings. Group is used when the source of the event is not important, and data in snapshots is enough for the hook. When binding with `group` is triggered with the event, the hook receives snapshots from all `kubernetes` bindings with the same `group` name.
```

File: HOOKS.md [hunk 9]
```diff
@@ -659,49 +674,67 @@ at 12:02, the hook will be executed with the following binding context:
...
 
-`executeHookOnSynchronization`, `executeHookOnEvent` and `keepFullObjectsInMemory` can be used with `group`.
+Adjacent tasks for `kubernetes` and `schedule` bindings with the same `group` and `queue` are "compacted" and hook is executed only once. So it is wise to use the same `queue` for all hooks in a group. This "compaction" mechanism is not available for `kubernetesValidating` and `kubernetesCustomResourceConversion` bindings as they're not queued.
```
Comment (nabokihms) [inline HOOKS.md:679]:
```suggestion
Adjacent tasks for `kubernetes` and `schedule` bindings with the same `group` and `queue` are "compacted", and the hook is executed only once. So it is wise to use the same `queue` for all hooks in a group. This "compaction" mechanism is not available for `kubernetesValidating` and `kubernetesCustomResourceConversion` bindings as they're not queued.
```

**Review by nabokihms — APPROVED** [review]
I have added some suggestion for the docs changes, but anyway, looks good and ready to merge.

File: pkg/kube_events_manager/kube_events_manager_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply pkg/kube_events_manager/kube_events_manager_test.go:340]: Could you also delete a commented line?

File: pkg/kube_events_manager/kube_events_manager_test.go [hunk 10]
```diff
@@ -336,12 +332,12 @@ func Test_FakeClient_CatchUpdates(t *testing.T) {
...
 
 	fmt.Printf("mgr Started\n")
```
Comment (diafour) [inline pkg/kube_events_manager/kube_events_manager_test.go:343]:
```suggestion
```

File: pkg/kube_events_manager/kube_events_manager_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/kube_events_manager/kube_events_manager_test.go:340]: Ouch! Missed this one.

**Pull request merged**